File: apps/desktop/src/settings/desktopState.ts
import { invokeNativeCommand } from "../native/commands";
import { isRecord } from "@thinkbrain/core";

export const DESKTOP_STATE_VERSION = 5;
export const DESKTOP_STATE_KEY = "desktopState";
export const MAX_RECENT_WORKSPACES = 12;
export const MIN_PANEL_WIDTH = 224;
export const MAX_PANEL_WIDTH = 480;
export const DEFAULT_LEFT_PANEL_WIDTH = 288;
export const DEFAULT_RIGHT_PANEL_WIDTH = 320;

/** Serializable tab metadata persisted across restarts. */
export interface PersistedTab {
  readonly id: string;
  readonly title: string;
  readonly kind: string;
  readonly rootPath?: string;
  readonly relativePath?: string;
}

/**
 * Per-workspace, per-view collapsed group keys (D53).
 *
 * What a user collapsed in a panel is not a preference they configured, so it
 * lives here rather than in their settings — that document now takes
 * conflict-safe writes, and churning it on every toggle would collide with the
 * writes that are preferences. Nor in the vault: it describes this machine.
 */
export type WorkspaceViews = Readonly<
  Record<string, Readonly<Record<string, readonly string[]>>>
>;

/** One view's collapsed groups, in one workspace. */
export interface CollapsedGroupsUpdate {
  readonly workspacePath: string;
  readonly viewId: string;
  readonly collapsed: readonly string[];
}

export interface DesktopState {
  readonly version: typeof DESKTOP_STATE_VERSION;
  readonly lastWorkspacePath: string | null;
  readonly recentWorkspacePaths: readonly string[];
  readonly explorerOpen: boolean;
  readonly leftPanelWidth: number;
  readonly rightPanelWidth: number;
  readonly bottomPanelOpen: boolean;
  readonly developmentExtensionDirectories: readonly string[];
  readonly openTabs: readonly PersistedTab[];
  readonly activeTabId: string | null;
  readonly workspaceViews: WorkspaceViews;
}

export interface DesktopStateUpdate {
  readonly lastWorkspacePath?: string | null;
  readonly recentWorkspacePaths?: readonly string[];
  readonly explorerOpen?: boolean;
  readonly leftPanelWidth?: number;
  readonly rightPanelWidth?: number;
  readonly bottomPanelOpen?: boolean;
  readonly developmentExtensionDirectories?: readonly string[];
  readonly openTabs?: readonly PersistedTab[];
  readonly activeTabId?: string | null;
  /**
   * Targeted rather than the whole map, because two windows on two vaults write
   * this without knowing about each other: sending the entire map would make
   * each of them overwrite what the other had just recorded.
   */
  readonly collapsedGroups?: CollapsedGroupsUpdate;
}

export interface DesktopStateGateway {
  readAppSettings(): Promise<string | null>;
  /**
   * `expected` is the document this write's `contents` were revised from —
   * `null` when none existed. `write_app_settings` refuses the write if that
   * is no longer what is on disk (see `appSettingsFile.ts`), so the fallback
   * path below must send what it actually read rather than nothing.
   */
  writeAppSettings(contents: string, expected: string | null): Promise<void>;
  updateDesktopState?(update: DesktopStateUpdate): Promise<string>;
}

export const DEFAULT_DESKTOP_STATE: DesktopState = Object.freeze({
  version: DESKTOP_STATE_VERSION,
  lastWorkspacePath: null,
  recentWorkspacePaths: [],
  explorerOpen: true,
  leftPanelWidth: DEFAULT_LEFT_PANEL_WIDTH,
  rightPanelWidth: DEFAULT_RIGHT_PANEL_WIDTH,
  bottomPanelOpen: false,
  developmentExtensionDirectories: [],
  openTabs: [],
  activeTabId: null,
  workspaceViews: {}
});

const nativeDesktopStateGateway: DesktopStateGateway = {
  readAppSettings: () => invokeNativeCommand("read_app_settings"),
  async writeAppSettings(contents, expected) {
    await invokeNativeCommand("write_app_settings", { contents, expected });
  },
  updateDesktopState(update) {
    return invokeNativeCommand("update_desktop_state", { update });
  }
};

/**
 * Reads desktop-only shell state from the shared app-settings document.
 * Invalid documents and unsupported state versions fall back to safe defaults.
 */
export async function loadDesktopState(
  gateway: DesktopStateGateway = nativeDesktopStateGateway
): Promise<DesktopState> {
  return parseDesktopState(await gateway.readAppSettings());
}

let fallbackUpdateQueue: Promise<unknown> = Promise.resolve();

/**
 * Updates desktop-only state without rewriting theme, extension, or other app
 * settings. The write always uses the current desktop-state schema.
 */
export async function saveDesktopState(
  update: DesktopStateUpdate,
  gateway: DesktopStateGateway = nativeDesktopStateGateway
): Promise<DesktopState> {
  if (gateway.updateDesktopState) {
    return parseDesktopState(await gateway.updateDesktopState(update));
  }

  console.warn(
    "Using fallback desktop state update. " +
    "This may cause race conditions if multiple windows modify settings concurrently."
  );

  // This does not go through `appSettingsFile.ts`'s retrying chain: that chain
  // exists for the settings store, which is a real writer in the shipped app,
  // races with `update_desktop_state` on every save, and is exercised by every
  // build. This branch only runs for a caller-supplied gateway that omits
  // `updateDesktopState` — never `nativeDesktopStateGateway`, which always
  // defines it — so it has no writer to race against in practice. Sending
  // `expected` keeps a theoretical race a loud failure instead of a silent
  // overwrite; a retry loop here would be safety net for a path nothing takes.
  const performUpdate = async () => {
    const rawAppSettings = await gateway.readAppSettings();
    const appSettings = parseAppSettingsRecord(rawAppSettings);
    const current = readDesktopState(appSettings);
    const next = applyDesktopStateUpdate(current, update);

    appSettings[DESKTOP_STATE_KEY] = next;
    delete appSettings.lastWorkspacePath;
    delete appSettings.explorerOpen;

    await gateway.writeAppSettings(serializeAppSettingsRecord(appSettings), rawAppSettings);
    return next;
  };

  const result = fallbackUpdateQueue.then(performUpdate, performUpdate);
  fallbackUpdateQueue = result.catch(() => {});
  return result;
}

/**
 * Parses either the current nested schema or the prior flat desktop fields.
 * This is exported to keep migration and malformed-document behavior testable.
 */
export function parseDesktopState(contents: string | null): DesktopState {
  return readDesktopState(parseAppSettingsRecord(contents));
}

function parseAppSettingsRecord(contents: string | null): Record<string, unknown> {
  if (contents === null) {
    return {};
  }

  try {
    const parsed: unknown = JSON.parse(contents);
    return isRecord(parsed) ? { ...parsed } : {};
  } catch {
    return {};
  }
}

function readDesktopState(appSettings: Readonly<Record<string, unknown>>): DesktopState {
  const storedState = appSettings[DESKTOP_STATE_KEY];

  if (isRecord(storedState)) {
    return readVersionedDesktopState(storedState);
  }

  return createDesktopState({
    lastWorkspacePath: appSettings.lastWorkspacePath,
    explorerOpen: appSettings.explorerOpen
  });
}

function readVersionedDesktopState(storedState: Readonly<Record<string, unknown>>): DesktopState {
  const version = storedState.version;

  // Every schema is readable in both directions — each version only ever added
  // fields, so a missing one falls back to its default and one this build has
  // not heard of is simply not read. A *newer* document is the one that used to
  // cost the most: running a newer build and then an older one is a branch
  // switch, and rejecting the document lost the workspace, the tabs and the
  // layout in one go. Only a version that is not a version at all is grounds to
  // give up, because then nothing can be said about the rest of the record.
  if (version !== undefined && !(typeof version === "number" && Number.isInteger(version) && version >= 0)) {
    return DEFAULT_DESKTOP_STATE;
  }

  return createDesktopState(storedState);
}

function applyDesktopStateUpdate(
  state: DesktopState,
  update: DesktopStateUpdate
): DesktopState {
  return createDesktopState({
    lastWorkspacePath:
      update.lastWorkspacePath === undefined
        ? state.lastWorkspacePath
        : update.lastWorkspacePath,
    recentWorkspacePaths:
      update.recentWorkspacePaths === undefined
        ? promoteRecentWorkspace(state.recentWorkspacePaths, update.lastWorkspacePath)
        : mergeRecentWorkspacePaths(
            state.recentWorkspacePaths,
            normalizeWorkspacePaths(update.recentWorkspacePaths)
          ),
    explorerOpen: update.explorerOpen === undefined ? state.explorerOpen : update.explorerOpen,
    leftPanelWidth:
      update.leftPanelWidth === undefined ? state.leftPanelWidth : update.leftPanelWidth,
    rightPanelWidth:
      update.rightPanelWidth === undefined ? state.rightPanelWidth : update.rightPanelWidth,
    bottomPanelOpen:
      update.bottomPanelOpen === undefined ? state.bottomPanelOpen : update.bottomPanelOpen,
    developmentExtensionDirectories:
      update.developmentExtensionDirectories === undefined
        ? state.developmentExtensionDirectories
        : update.developmentExtensionDirectories,
    openTabs: update.openTabs === undefined ? state.openTabs : update.openTabs,
    activeTabId: update.activeTabId === undefined ? state.activeTabId : update.activeTabId,
    workspaceViews: applyCollapsedGroups(state.workspaceViews, update.collapsedGroups)
  });
}

/**
 * Records one view's collapsed groups.
 *
 * Unlike the native path this does not prune workspaces the app has forgotten:
 * this branch only runs for a caller-supplied gateway without
 * `updateDesktopState`, which the shipped app never takes.
 */
function applyCollapsedGroups(
  views: WorkspaceViews,
  update: CollapsedGroupsUpdate | undefined
): WorkspaceViews {
  if (update === undefined) return views;
  return {
    ...views,
    [update.workspacePath]: {
      ...views[update.workspacePath],
      [update.viewId]: update.collapsed
    }
  };
}

function createDesktopState(value: Readonly<Record<string, unknown>>): DesktopState {
  const lastWorkspacePath = readNonEmptyString(value.lastWorkspacePath);
  return {
    version: DESKTOP_STATE_VERSION,
    lastWorkspacePath,
    recentWorkspacePaths: normalizeWorkspacePaths(value.recentWorkspacePaths, lastWorkspacePath),
    explorerOpen:
      typeof value.explorerOpen === "boolean"
        ? value.explorerOpen
        : DEFAULT_DESKTOP_STATE.explorerOpen,
    leftPanelWidth: readPanelWidth(value.leftPanelWidth, DEFAULT_LEFT_PANEL_WIDTH),
    rightPanelWidth: readPanelWidth(value.rightPanelWidth, DEFAULT_RIGHT_PANEL_WIDTH),
    bottomPanelOpen:
      typeof value.bottomPanelOpen === "boolean"
        ? value.bottomPanelOpen
        : DEFAULT_DESKTOP_STATE.bottomPanelOpen,
    developmentExtensionDirectories: normalizeExtensionDirectories(
      value.developmentExtensionDirectories
    ),
    openTabs: readPersistedTabs(value.openTabs),
    activeTabId: readNonEmptyString(value.activeTabId),
    workspaceViews: readWorkspaceViews(value.workspaceViews)
  };
}

/**
 * Reads the stored collapsed groups, skipping anything malformed.
 *
 * A hand-edited or truncated document must cost the user a group that reopened,
 * never a panel that will not draw.
 */
function readWorkspaceViews(value: unknown): WorkspaceViews {
  if (!isRecord(value)) return {};
  const views: Record<string, Record<string, readonly string[]>> = {};
  for (const [workspacePath, stored] of Object.entries(value)) {
    if (!isRecord(stored)) continue;
    const byView: Record<string, readonly string[]> = {};
    for (const [viewId, collapsed] of Object.entries(stored)) {
      if (!Array.isArray(collapsed)) continue;
      byView[viewId] = collapsed.filter((key): key is string => typeof key === "string");
    }
    views[workspacePath] = byView;
  }
  return views;
}

/**
 * The groups collapsed in one view of one workspace, or none.
 *
 * A workspace nothing was stored for reads the same as one stored empty: every
 * group open, which is the right answer for a panel opened for the first time.
 */
export function collapsedGroups(
  state: DesktopState,
  workspacePath: string | null,
  viewId: string
): readonly string[] {
  if (workspacePath === null) return [];
  return state.workspaceViews[workspacePath]?.[viewId] ?? [];
}

/**
 * Deduplicates extension directories without resolving them: a directory that
 * is temporarily missing must stay stored so the user can fix it.
 */
function normalizeExtensionDirectories(value: unknown): readonly string[] {
  const directories = Array.isArray(value)
    ? value.filter((entry): entry is string => typeof entry === "string" && entry.length > 0)
    : [];
  return [...new Set(directories)];
}

/** Reads and validates the persisted tab list from a desktop-state record. */
function readPersistedTabs(value: unknown): readonly PersistedTab[] {
  if (!Array.isArray(value)) return [];
  return value
    .filter((entry): entry is Record<string, unknown> => isRecord(entry))
    .map((entry) => ({
      id: typeof entry.id === "string" ? entry.id : "",
      title: typeof entry.title === "string" ? entry.title : "",
      kind: typeof entry.kind === "string" ? entry.kind : "editor",
      ...(typeof entry.rootPath === "string" && entry.rootPath ? { rootPath: entry.rootPath } : {}),
      ...(typeof entry.relativePath === "string" && entry.relativePath
        ? { relativePath: entry.relativePath }
        : {})
    }))
    .filter((tab) => tab.id.length > 0);
}

/** Reads a non-empty string value, or null when absent/invalid. */
function readNonEmptyString(value: unknown): string | null {
  return typeof value === "string" && value.length > 0 ? value : null;
}

/** Clamps a dock width to the range that keeps the shell usable. */
export function clampPanelWidth(width: number): number {
  return Math.max(MIN_PANEL_WIDTH, Math.min(MAX_PANEL_WIDTH, width));
}

function readPanelWidth(value: unknown, fallback: number): number {
  return typeof value === "number" && Number.isFinite(value)
    ? clampPanelWidth(value)
    : fallback;
}

function normalizeWorkspacePaths(value: unknown, fallback?: string | null): readonly string[] {
  const paths = Array.isArray(value)
    ? value.filter((path): path is string => typeof path === "string" && path.length > 0)
    : [];
  return promoteRecentWorkspace(paths, fallback);
}

export function promoteRecentWorkspace(paths: readonly string[], path: string | null | undefined): readonly string[] {
  const unique = [...new Set(path ? [path, ...paths] : paths)];
  return unique.slice(0, MAX_RECENT_WORKSPACES);
}

/**
 * Merges an explicitly provided recent-workspace list with the stored one
 * instead of replacing it outright. A caller's list can be stale relative to
 * another window's concurrent write, so this preserves entries neither side
 * sent explicitly (mirrors Rust's `merge_recent_workspace_paths`).
 */
function mergeRecentWorkspacePaths(
  current: readonly string[],
  incoming: readonly string[]
): readonly string[] {
  return promoteRecentWorkspace([...incoming, ...current], undefined);
}

function serializeAppSettingsRecord(appSettings: Readonly<Record<string, unknown>>): string {
  return `${JSON.stringify(appSettings, null, 2)}\n`;
}

File: apps/desktop/src-tauri/src/tests/desktop_state.rs
//! Desktop-state tests: MRU merging, extension directories, schema version
//! tolerance (newer/older builds), active-tab clearing semantics, and the
//! per-workspace `workspaceViews` (collapsed groups) and `workspaceTabs` maps.

use crate::commands::settings::*;
use crate::commands::workspace::resolve_workspace_root;
use serde_json::Value;
use std::fs;

use super::temp_test_dir;

#[test]
fn desktop_state_update_merges_concurrent_mrus_and_preserves_app_settings() {
    let temp_dir = temp_test_dir("state_merge");
    let one = temp_dir.join("one");
    let two = temp_dir.join("two");
    let legacy = temp_dir.join("legacy");
    std::fs::create_dir_all(&one).unwrap();
    std::fs::create_dir_all(&two).unwrap();
    std::fs::create_dir_all(&legacy).unwrap();

    let one_path = resolve_workspace_root(&one.to_string_lossy())
        .unwrap()
        .to_string_lossy()
        .to_string();
    let two_path = resolve_workspace_root(&two.to_string_lossy())
        .unwrap()
        .to_string_lossy()
        .to_string();
    let legacy_path = resolve_workspace_root(&legacy.to_string_lossy())
        .unwrap()
        .to_string_lossy()
        .to_string();

    let first_json = serde_json::json!({
        "theme": "dark",
        "extensionSettings": { "timer": { "enabled": true } },
        "lastWorkspacePath": legacy_path,
        "explorerOpen": false
    });

    let first = update_desktop_state_contents(
        Some(&first_json.to_string()),
        DesktopStateUpdate {
            last_workspace_path: Some(Some(one_path.clone())),
            recent_workspace_paths: Some(vec![one_path.clone(), legacy_path.clone()]),
            left_panel_width: Some(352.0),
            bottom_panel_open: Some(true),
            ..Default::default()
        },
    )
    .expect("first desktop-state update succeeds");

    let second = update_desktop_state_contents(
        Some(&first),
        DesktopStateUpdate {
            recent_workspace_paths: Some(vec![two_path.clone(), legacy_path.clone()]),
            explorer_open: Some(true),
            right_panel_width: Some(512.0),
            ..Default::default()
        },
    )
    .expect("second desktop-state update succeeds");

    let settings: Value = serde_json::from_str(&second).expect("serialized settings are valid");
    assert_eq!(settings["theme"], serde_json::json!("dark"));
    assert_eq!(
        settings["extensionSettings"]["timer"]["enabled"],
        serde_json::json!(true)
    );
    // Legacy flat-schema keys are no longer migrated/removed — they are
    // preserved as unrelated app settings (DESKTOP_STATE_VERSION >= 5).
    assert_eq!(
        settings["lastWorkspacePath"],
        serde_json::json!(legacy_path)
    );
    assert_eq!(settings["explorerOpen"], serde_json::json!(false));
    assert_eq!(
        settings["desktopState"],
        serde_json::json!({
            "version": 5,
            "lastWorkspacePath": one_path,
            "recentWorkspacePaths": [two_path, legacy_path, one_path],
            "workspaceViews": {},
            "workspaceTabs": {},
            "explorerOpen": true,
            "leftPanelWidth": 352.0,
            "rightPanelWidth": 480.0,
            "bottomPanelOpen": true,
            "developmentExtensionDirectories": [],
            "openTabs": [],
            "activeTabId": null
        })
    );
}

#[test]
fn desktop_state_persists_development_extension_directories_verbatim() {
    // Directories are stored as given — not canonicalized — so a directory
    // that is temporarily missing stays in the list instead of vanishing.
    let stored = update_desktop_state_contents(
        None,
        DesktopStateUpdate {
            development_extension_directories: Some(vec![
                "/ext/one".to_string(),
                "".to_string(),
                "/ext/two".to_string(),
                "/ext/one".to_string(),
            ]),
            ..Default::default()
        },
    )
    .expect("desktop-state update succeeds");

    let settings: Value = serde_json::from_str(&stored).expect("serialized settings are valid");
    assert_eq!(
        settings["desktopState"]["developmentExtensionDirectories"],
        serde_json::json!(["/ext/one", "/ext/two"])
    );

    // An update that does not mention the field keeps the stored list.
    let unchanged = update_desktop_state_contents(
        Some(&stored),
        DesktopStateUpdate {
            explorer_open: Some(true),
            ..Default::default()
        },
    )
    .expect("unrelated desktop-state update succeeds");

    let settings: Value = serde_json::from_str(&unchanged).expect("serialized settings are valid");
    assert_eq!(
        settings["desktopState"]["developmentExtensionDirectories"],
        serde_json::json!(["/ext/one", "/ext/two"])
    );
}

#[test]
fn desktop_state_without_extension_directories_defaults_to_empty() {
    let existing = serde_json::json!({
        "desktopState": { "version": 3, "explorerOpen": true }
    });

    let updated =
        update_desktop_state_contents(Some(&existing.to_string()), DesktopStateUpdate::default())
            .expect("desktop-state update succeeds");

    let settings: Value = serde_json::from_str(&updated).expect("serialized settings are valid");
    assert_eq!(
        settings["desktopState"]["developmentExtensionDirectories"],
        serde_json::json!([])
    );
}

/// A branch switch runs a newer build and then an older one. Treating the newer
/// build's document as unreadable threw away the workspace, the open tabs and
/// the panel layout in one write; every schema here is additive, so the older
/// build can read all of it but the fields it has never heard of.
#[test]
fn desktop_state_from_a_newer_build_is_read_rather_than_discarded() {
    let existing = serde_json::json!({
        "desktopState": {
            "version": 99,
            "explorerOpen": false,
            "leftPanelWidth": 352.0,
            "somethingLaterAdded": "not understood here"
        }
    });

    let updated =
        update_desktop_state_contents(Some(&existing.to_string()), DesktopStateUpdate::default())
            .expect("desktop-state update succeeds");

    let settings: Value = serde_json::from_str(&updated).expect("serialized settings are valid");
    // Both differ from their defaults, so reading them back is the proof.
    assert_eq!(settings["desktopState"]["explorerOpen"], false);
    assert_eq!(settings["desktopState"]["leftPanelWidth"], 352.0);
}

/// Reading a newer document is only half of it: this build then writes the
/// document back, and anything it rebuilt from scratch would drop whatever the
/// newer build had added. Carrying the unknown fields through — and leaving the
/// version where it was — makes the round trip lossless, so switching branches
/// costs nothing at all rather than costing the newest feature's state.
#[test]
fn desktop_state_from_a_newer_build_survives_a_write_by_this_one() {
    let existing = serde_json::json!({
        "desktopState": {
            "version": 99,
            "explorerOpen": false,
            "somethingLaterAdded": { "kept": true }
        }
    });

    let updated = update_desktop_state_contents(
        Some(&existing.to_string()),
        DesktopStateUpdate {
            bottom_panel_open: Some(true),
            ..Default::default()
        },
    )
    .expect("desktop-state update succeeds");

    let settings: Value = serde_json::from_str(&updated).expect("serialized settings are valid");
    assert_eq!(
        settings["desktopState"]["somethingLaterAdded"],
        serde_json::json!({ "kept": true })
    );
    assert_eq!(settings["desktopState"]["version"], 99);
    assert_eq!(settings["desktopState"]["bottomPanelOpen"], true);
    assert_eq!(settings["desktopState"]["explorerOpen"], false);
}

#[test]
fn desktop_state_with_a_version_that_is_not_a_version_falls_back_to_defaults() {
    let existing = serde_json::json!({
        "desktopState": { "version": "five", "explorerOpen": false, "leftPanelWidth": 352.0 }
    });

    let updated =
        update_desktop_state_contents(Some(&existing.to_string()), DesktopStateUpdate::default())
            .expect("desktop-state update succeeds");

    let settings: Value = serde_json::from_str(&updated).expect("serialized settings are valid");
    assert_eq!(settings["desktopState"]["explorerOpen"], true);
    assert_eq!(settings["desktopState"]["leftPanelWidth"], 288.0);
}

#[test]
fn desktop_state_active_tab_id_explicit_null_clears_instead_of_restoring_current() {
    // Mirrors `last_workspace_path`'s `Some(None)`-clears semantics: an
    // explicit null must clear the active tab rather than keep the old one.
    let stored = update_desktop_state_contents(
        None,
        DesktopStateUpdate {
            active_tab_id: Some(Some("tab-1".to_string())),
            ..Default::default()
        },
    )
    .expect("desktop-state update succeeds");

    let settings: Value = serde_json::from_str(&stored).expect("serialized settings are valid");
    assert_eq!(
        settings["desktopState"]["activeTabId"],
        serde_json::json!("tab-1")
    );

    let cleared = update_desktop_state_contents(
        Some(&stored),
        DesktopStateUpdate {
            active_tab_id: Some(None),
            ..Default::default()
        },
    )
    .expect("desktop-state update succeeds");

    let settings: Value = serde_json::from_str(&cleared).expect("serialized settings are valid");
    assert_eq!(settings["desktopState"]["activeTabId"], Value::Null);

    // An update that omits the field entirely keeps the current value.
    let restored = update_desktop_state_contents(Some(&stored), DesktopStateUpdate::default())
        .expect("desktop-state update succeeds");

    let settings: Value = serde_json::from_str(&restored).expect("serialized settings are valid");
    assert_eq!(
        settings["desktopState"]["activeTabId"],
        serde_json::json!("tab-1")
    );
}

/// D53: a group the user collapsed stays collapsed, per workspace, in desktop
/// state rather than in their settings or in the vault.
#[test]
fn collapsed_groups_are_kept_per_workspace_and_per_view() {
    // Real directories, because a workspace path is canonicalized before it is
    // stored — an imaginary one would be dropped and prove nothing.
    let first_root = temp_test_dir("collapse_first");
    let second_root = temp_test_dir("collapse_second");
    let first_path = first_root.to_string_lossy().to_string();
    let second_path = second_root.to_string_lossy().to_string();

    let collapse = |contents: Option<&str>, workspace: &str, view: &str, keys: Vec<String>| {
        update_desktop_state_contents(
            contents,
            DesktopStateUpdate {
                last_workspace_path: Some(Some(workspace.to_string())),
                collapsed_groups: Some(CollapsedGroupsUpdate {
                    workspace_path: workspace.to_string(),
                    view_id: view.to_string(),
                    collapsed: keys,
                }),
                ..Default::default()
            },
        )
        .expect("desktop-state update succeeds")
    };

    let stored = collapse(
        None,
        &first_path,
        "journal",
        vec!["2026".to_string(), "2026-08".to_string()],
    );
    // A second vault must not overwrite the first: two windows write this field
    // without knowing about each other.
    let stored = collapse(
        Some(&stored),
        &second_path,
        "journal",
        vec!["2025".to_string()],
    );

    // A second view of the same vault sits beside the first rather than
    // replacing it: the explorer tree has the same problem and will want a row
    // here, and one write must not take the other's.
    let stored = collapse(
        Some(&stored),
        &first_path,
        "explorer",
        vec!["notes".to_string()],
    );

    let settings: Value = serde_json::from_str(&stored).expect("serialized settings are valid");
    let views = &settings["desktopState"]["workspaceViews"];
    assert_eq!(
        views[&first_path]["journal"],
        serde_json::json!(["2026", "2026-08"])
    );
    assert_eq!(views[&first_path]["explorer"], serde_json::json!(["notes"]));
    assert_eq!(views[&second_path]["journal"], serde_json::json!(["2025"]));

    // Reopening a group writes a shorter list, and an empty one is a real answer
    // rather than an absent one — every group is open.
    let reopened = collapse(Some(&stored), &first_path, "journal", Vec::new());
    let settings: Value = serde_json::from_str(&reopened).expect("serialized settings are valid");
    assert_eq!(
        settings["desktopState"]["workspaceViews"][&first_path]["journal"],
        serde_json::json!([])
    );

    fs::remove_dir_all(&first_root).ok();
    fs::remove_dir_all(&second_root).ok();
}

/// Two windows on two vaults must not show each other's tabs.
///
/// Tabs used to be one flat list in a document every window shares, so the last
/// window to touch a tab overwrote the other's list, and on the next launch both
/// restored the same tabs — pointing at whichever vault had written them. They
/// are keyed by workspace for the same reason the collapsed groups are, and
/// written the same targeted way: sending the whole map would put back exactly
/// the overwriting this removes.
#[test]
fn tabs_are_kept_per_workspace() {
    let first_root = temp_test_dir("tabs_first");
    let second_root = temp_test_dir("tabs_second");
    let first_path = first_root.to_string_lossy().to_string();
    let second_path = second_root.to_string_lossy().to_string();

    let open = |contents: Option<&str>, workspace: &str, note: &str| {
        update_desktop_state_contents(
            contents,
            DesktopStateUpdate {
                last_workspace_path: Some(Some(workspace.to_string())),
                workspace_tabs: Some(WorkspaceTabsUpdate {
                    workspace_path: workspace.to_string(),
                    open_tabs: vec![PersistedTab {
                        id: format!("editor:{workspace}:{note}"),
                        title: note.to_string(),
                        kind: "editor".to_string(),
                        root_path: Some(workspace.to_string()),
                        relative_path: Some(note.to_string()),
                    }],
                    active_tab_id: Some(format!("editor:{workspace}:{note}")),
                }),
                ..Default::default()
            },
        )
        .expect("desktop-state update succeeds")
    };

    let stored = open(None, &first_path, "one.md");
    // The second window must leave the first window's tabs alone.
    let stored = open(Some(&stored), &second_path, "two.md");

    let settings: Value = serde_json::from_str(&stored).expect("serialized settings are valid");
    let tabs = &settings["desktopState"]["workspaceTabs"];

    assert_eq!(tabs[&first_path]["openTabs"][0]["relativePath"], "one.md");
    assert_eq!(tabs[&second_path]["openTabs"][0]["relativePath"], "two.md");
    assert_eq!(
        tabs[&first_path]["openTabs"].as_array().map(Vec::len),
        Some(1),
        "the second window added its tab to the first window's list"
    );
    assert_eq!(
        tabs[&first_path]["activeTabId"],
        format!("editor:{first_path}:one.md")
    );

    // Closing every tab is a real answer, not an absent one.
    let emptied = update_desktop_state_contents(
        Some(&stored),
        DesktopStateUpdate {
            last_workspace_path: Some(Some(first_path.clone())),
            workspace_tabs: Some(WorkspaceTabsUpdate {
                workspace_path: first_path.clone(),
                open_tabs: Vec::new(),
                active_tab_id: None,
            }),
            ..Default::default()
        },
    )
    .expect("desktop-state update succeeds");
    let settings: Value = serde_json::from_str(&emptied).expect("serialized settings are valid");
    assert_eq!(
        settings["desktopState"]["workspaceTabs"][&first_path]["openTabs"],
        serde_json::json!([])
    );
    // ...and it does not disturb the other vault.
    assert_eq!(
        settings["desktopState"]["workspaceTabs"][&second_path]["openTabs"][0]["relativePath"],
        "two.md"
    );

    fs::remove_dir_all(&first_root).ok();
    fs::remove_dir_all(&second_root).ok();
}

/// The stored views follow the recent-workspace list rather than carrying a
/// bound of their own, so a vault the app has forgotten stops costing anything.
#[test]
fn collapsed_groups_are_dropped_for_a_workspace_no_longer_remembered() {
    let root = temp_test_dir("collapse_forgotten");
    let path = root.to_string_lossy().to_string();

    let stored = update_desktop_state_contents(
        None,
        DesktopStateUpdate {
            last_workspace_path: Some(Some(path.clone())),
            collapsed_groups: Some(CollapsedGroupsUpdate {
                workspace_path: path.clone(),
                view_id: "journal".to_string(),
                collapsed: vec!["2026".to_string()],
            }),
            ..Default::default()
        },
    )
    .expect("desktop-state update succeeds");

    let settings: Value = serde_json::from_str(&stored).expect("serialized settings are valid");
    assert!(settings["desktopState"]["workspaceViews"][&path].is_object());

    // Forgetting the vault takes what was collapsed in it: there is no panel
    // left to restore, and one policy is easier to reason about than two.
    fs::remove_dir_all(&root).ok();
    let forgotten = update_desktop_state_contents(
        Some(&stored),
        DesktopStateUpdate {
            last_workspace_path: Some(None),
            recent_workspace_paths: Some(Vec::new()),
            ..Default::default()
        },
    )
    .expect("update succeeds");

    let settings: Value = serde_json::from_str(&forgotten).expect("serialized settings are valid");
    assert_eq!(
        settings["desktopState"]["workspaceViews"],
        serde_json::json!({})
    );
}
